modules/collect/src/main/java/com/opengamma/strata/collect/io/CsvIterator.java

```diff
@@ -0,0 +1,281 @@
+ * <pre>
+ *  try (CsvIterable iterable = CsvIterable.of(source, true)) {
+ *    // use the CsvIterable
+ *  }
+ * </pre>
+ * One way to use the iterable is with the for-each loop, using a lambda to adapt {@code Iterator} to {@code Iterable}:
+ * <pre>
+ *  try (CsvIterable iterable = CsvIterable.of(source, true)) {
+ *    for (CsvRow row : () -> iterable) {
+ *      // process the row
+ *    }
+ *  }
+ * </pre>
+ * This class also allows the headers to be obtained without reading the whole CSV file:
+ * <pre>
+ *  try (CsvIterable iterable = CsvIterable.of(source, true)) {
+ *    ImmutableList{@literal <String>} headers = iterable.headers();
+ *  }
+ * </pre>
+ */
+public final class CsvIterator implements AutoCloseable, PeekingIterator<CsvRow> {
+
+  /**
+   * The buffered reader.
+   */
+  private final BufferedReader reader;
+  /**
+   * The separator
+   */
+  private final char separator;
+  /**
+   * The header row, ordered as the headers appear in the file.
+   */
+  private final ImmutableList<String> headers;
+  /**
+   * The header map, transformed for case-insensitive searching.
+   */
+  private final ImmutableMap<String, Integer> searchHeaders;
+  /**
+   * The next row.
+   */
+  private CsvRow nextRow;
+
+  //------------------------------------------------------------------------
+  /**
+   * Parses the specified source as a CSV file.
+   * 
+   * @param source  the CSV file resource
+   * @param headerRow  whether the source has a header row, an empty source must still contain the header
+   * @return the CSV file
+   * @throws UncheckedIOException if an IO exception occurs
+   * @throws IllegalArgumentException if the file cannot be parsed
+   */
+  public static CsvIterator of(CharSource source, boolean headerRow) {
+    return of(source, headerRow, ',');
+  }
+
+  /**
+   * Parses the specified source as a CSV file where the separator is specified and might not be a comma.
+   * <p>
+   * This overload allows the separator to be controlled.
+   * For example, a tab-separated file is very similar to a CSV file, the only difference is the separator.
+   * 
+   * @param source  the file resource
+   * @param headerRow  whether the source has a header row, an empty source must still contain the header
+   * @param separator  the separator used to separate each field, typically a comma, but a tab is sometimes used
+   * @return the TSV file
+   * @throws UncheckedIOException if an IO exception occurs
+   * @throws IllegalArgumentException if the file cannot be parsed
+   */
+  public static CsvIterator of(CharSource source, boolean headerRow, char separator) {
+    ArgChecker.notNull(source, "source");
+
+    return Unchecked.wrap(() -> {
+      BufferedReader reader = source.openBufferedStream();
+      if (!headerRow) {
+        return new CsvIterator(reader, separator, ImmutableList.of(), ImmutableMap.of());
+      }
+      String line = reader.readLine();
```

Comment:
If this line or the following if-block throws an exception the reader will not be closed.